### ISSUE TITLE
destructor double-destroy due to break within loop at top-level

### ISSUE THREAD

**Issue by alaviss** [nim-works/nimskull#1010]
### Example

```nim
type
  X = object
    destroyed: bool

proc `=destroy`(x: var X) =
  doAssert not x.destroyed
  x.destroyed = true

for i in 1..2:
  if i == 2:
    break
  let c = X()
  echo c
```

### Actual Output

```
(destroyed: false)
iterators_1.nim(92)      test
test.nim(6)              =destroy
assertions.nim(38)       failedAssertImpl
assertions.nim(28)       raiseAssert
fatal.nim(48)            sysFatal
Error: unhandled exception: test.nim(6, 3) `not x.destroyed`  [AssertionDefect]
```

### Expected Output

```
(destroyed: false)
```

### Additional Information

If the block is in a `main()` proc, the double-free does not occur.

Some observation:

- The `c` local was not reset on loop start when the code is at top-level, but was reset when the code is in a proc.
- The variable `c` was promoted to global, even though it is bound by the loop's scope.

### References

* Found while testing `nimph`.

**Comment by zerbina** [comment]
The funamental problem is that the `injectdestructors` pass currently only wraps full scopes in a `try`/`finally` (which is done when there exists unstructured control-flow leaving the scope and at least one local requiring destruction is still alive).

To uphold the destruction order, all destructor calls (even for locals that aren't alive at any unstructured control-flow exits) are lifted into the finalizer, but this leads to unexpected behaviour:
```nim
proc main(cond: bool) =
  let a = X()
  if cond:
    return
  let b = X()
```
Here, the destructor for `b` is called even when the return is taken!

While the problem with the unexpected (and also wrong destruction) will go away once cleanup handling in the `injectdestructors` pass is reworked, the problem-as-reported can, in the meantime, be fixed by a change to how non-top-level globals are processed. I'll make a PR.

**Comment by alaviss** [comment]
Apparently this wasn't auto-closed by #1013, manually closing.
